### ISSUE TITLE
Fix S3242: Rule throws ArgumentException

### ISSUE THREAD

**Issue by antrv** [SonarSource/sonar-csharp#803]
![argumentexception](https://user-images.githubusercontent.com/1135617/31072477-c9c73268-a770-11e7-84c9-f2576b95b40c.png)

No code sample for reproducing this bug yet (I cannot provide original code).

**Comment by antrv** [comment]
I could not catch the descriptive message again, so I attached the screenshot.

**Comment by Evangelink** [comment]
Hi @antrv,
I was reviewing the code and just saw we build a dictionary with the name of parameters as key (which seems to match your exception). Could you confirm you have any method with multiple parameters with the same name? I will try to make some tests later on but for now we have a lot of things to deal with on SonarLint for VS.

**Comment by antrv** [comment]
yes, I have many methods with the same parameter names, only type parameters differ.
for example:
Method<T1, T2>(Tuple<T1, T2> tuple, Action<Tuple<T1, T2>> action)
Method<T1, T2, T3>(Tuple<T1, T2, T3> tuple, Action<Tuple<T1, T2, T3>> action)

**Comment by antrv** [comment]
and even (simplified)
Method<T1, T2, T3>(Tuple<Tuple<T1>, Tuple<T2, T3>> tuple, Action<Tuple<T1>, Tuple<T2, T3>> action)
and
Method<T1, T2, T3>(Tuple<Tuple<T1, T2>, Tuple<T3>> tuple, Action<Tuple<T1, T2>, Tuple<T3>> action)

**Comment by Evangelink** [comment]
Hi @antrv,
Sorry to get back to you so late but the only case where I can reproduce the error is actually when I create a method that doesn't compile with 2 or more parameters with the same name. For example:
```
private void Foo(IList<int> a, IList<string> a)
{
    a.ToList();
}
```

Does it match your use case or do you think you are having a different bug?

**Comment by antrv** [comment]
Hi @Evangelink ,
I think I have a different bug. My code compiles.
Please keep this bug open for a week or two. I will try to provide code sample for it.
Thank you

**Comment by Evangelink** [comment]
Will do :)

**Comment by Evangelink** [comment]
Hi @antrv,

We have received a similar bug through the [SonarLint Google Group](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/sonarlint/bWXR_YutZ3Q/qw5sKbk4BwAJ):
![image](https://user-images.githubusercontent.com/11340282/31876874-8d27dc3a-b7d4-11e7-945d-d16b5e961499.png)
So it seems like the problem is when adding twice the same usage (the rule looks at all usages to know the least derived type that can be used).

**Comment by Evangelink** [comment]
Fixed by PR #835
